### PR TITLE
[CLI] Accept also equal sign "=" for ``--libraries``

### PR DESCRIPTION
**Pull request by ityuhui**
To fix the issue https://github.com/ethereum/solidity/issues/10305

### CODE REVIEW

File: solc/CommandLineInterface.cpp [hunk 1]
```diff
@@ -664,31 +664,40 @@ bool CommandLineInterface::parseLibraryOption(string const& _input)
 	for (string const& lib: libraries)
 		if (!lib.empty())
 		{
-			//search for last colon in string as our binaries output placeholders in the form of file:Name
-			//so we need to search for the second `:` in the string
-			auto colon = lib.rfind(':');
-			if (colon == string::npos)
+			//search for equal sign in string as our binaries output placeholders in the form of file=Name
+			//so we need to search for `=` in the string
+			auto equalSign = lib.find('=');
```
Comment (leonardoalt) [inline solc/CommandLineInterface.cpp]: Why the change from `rfind` to `find`? I guess before it was `rfind` exactly because of the double `:` problem. Maybe we should `solAssert(lib.rfind('=') == lib.find('='), "");`?

File: solc/CommandLineInterface.cpp [hunk 2]
```diff
@@ -664,31 +664,40 @@ bool CommandLineInterface::parseLibraryOption(string const& _input)
...
 				serr() << "Address specified more than once for library \"" << libName << "\"." << endl;
 				return false;
 			}
 
-			string addrString(lib.begin() + static_cast<ptrdiff_t>(colon) + 1, lib.end());
+			string addrString(lib.begin() + static_cast<ptrdiff_t>(equalSign) + 1, lib.end());
 			boost::trim(addrString);
-			if (addrString.substr(0, 2) == "0x")
+			if (addrString.substr(0, 2) == "0x" || addrString.substr(0, 2) == "0X")
+			{
```
Comment (leonardoalt) [inline solc/CommandLineInterface.cpp:702]: Can remove these brackets

File: solc/CommandLineInterface.cpp [hunk 3]
```diff
@@ -664,31 +664,40 @@ bool CommandLineInterface::parseLibraryOption(string const& _input)
...
 				addrString = addrString.substr(2);
+			}
+			else
+			{
+				serr() << "The address "<< addrString << " is not prefixed with \"0x\"." << endl;
```
Comment (leonardoalt) [inline solc/CommandLineInterface.cpp]:
```suggestion
				serr() << "The address " << addrString << " is not prefixed with \"0x\"." << endl;
```

**Review by leonardoalt — COMMENTED** [review]
Thanks for the PR!
Left a few minor comments but looks good overall.

**Comment by leonardoalt** [comment]
@axic @chriseth btw should this go to `breaking` instead?

**Comment by leonardoalt** [comment]
Please also add a Changelog entry regarding the CLI

File: solc/CommandLineInterface.cpp (hunk 1 above)
Comment (ityuhui) [reply solc/CommandLineInterface.cpp]: The reason I changed ```rfind``` to ```find``` is I think there is only one "="  in the string and ```find``` is clearer for developer.

But ```rfind``` is OK to me. I will rollback.

Do we still need ```solAssert(lib.rfind('=') == lib.find('='), "Only one equal sign \"=\" is allowed in the address string.");``` here ?

Comment (leonardoalt) [reply solc/CommandLineInterface.cpp]: I'm fine with `find` instead of `rfind`. Maybe I'm just being too paranoid with the assertion

**Comment by axic** [comment]
> btw should this go to breaking instead?

It is breaking, as the original issue was labeled. We could support both options in a non-breaking manner (and drop the old in the breaking release), if we want to go as far as that. Not sure it is worth it.

File: solc/CommandLineInterface.cpp (hunk 1 above)
Comment (ityuhui) [reply solc/CommandLineInterface.cpp]: Sure. I have added ```solAssert``` here.

File: solc/CommandLineInterface.cpp (hunk 2 above)
Comment (ityuhui) [reply solc/CommandLineInterface.cpp:702]: OK. Done

File: solc/CommandLineInterface.cpp (hunk 3 above)
Comment (ityuhui) [reply solc/CommandLineInterface.cpp]: Done

**Comment by ityuhui** [comment]
So does the new design support both ":" and "=" for  the separator ?

File: solc/CommandLineInterface.cpp [hunk 4]
```diff
@@ -664,31 +664,39 @@ bool CommandLineInterface::parseLibraryOption(string const& _input)
...
 			{
 				serr() << "Address specified more than once for library \"" << libName << "\"." << endl;
 				return false;
 			}
 
-			string addrString(lib.begin() + static_cast<ptrdiff_t>(colon) + 1, lib.end());
+			string addrString(lib.begin() + static_cast<ptrdiff_t>(equalSign) + 1, lib.end());
 			boost::trim(addrString);
-			if (addrString.substr(0, 2) == "0x")
+			if (addrString.substr(0, 2) == "0x" || addrString.substr(0, 2) == "0X")
```
Comment (chriseth) [inline solc/CommandLineInterface.cpp]: Is `0X` actually used?

**Comment by chriseth** [comment]
I think it would be better to support both `:` and `=` for now.

File: solc/CommandLineInterface.cpp (hunk 4 above)
Comment (ityuhui) [reply solc/CommandLineInterface.cpp]: I will remove this part for ```0X``` if solidity CLI does not allow the prefix.

**Comment by ityuhui** [comment]
> I think it would be better to support both `:` and `=` for now.

OK. I will update code to implement this.

**Comment by ityuhui** [comment]
Hi there

The code is updated to support both ``=`` and ``:``, could you please review again ? Thanks

File: Changelog.md [hunk 5]
```diff
@@ -13,6 +13,7 @@ Compiler Features:
  * SMTChecker: Support try/catch statements.
  * SMTChecker: Output internal and trusted external function calls in a counterexample's transaction trace.
  * SMTChecker: Synthesize untrusted functions called externally.
+ * Command Line Interface: Support equal sign "=" for ```--libraries``` option.
```
Comment (leonardoalt) [inline Changelog.md]:
```suggestion
 * Command Line Interface: Accept also equal sign "=" for ``--libraries``.
```

File: solc/CommandLineInterface.cpp [hunk 6]
```diff
@@ -664,34 +664,51 @@ bool CommandLineInterface::parseLibraryOption(string const& _input)
...
+				if (separator == string::npos)
+				{
+					serr() << "Equal sign separator missing in library address specifier \"" << lib << "\"" << endl;
+					return false;
+				}
+				else
+					isSeparatorEqualSign = false; // separator is colon
 			}
+			else
+				solAssert(lib.rfind('=') == lib.find('='), "Only one equal sign \"=\" is allowed in the address string \"" + lib + "\".");
```
Comment (leonardoalt) [inline solc/CommandLineInterface.cpp]: Nice

**Review by leonardoalt — COMMENTED** [review]
Minor comment about the changelog entry, otherwise looks good!

File: Changelog.md (hunk 5 above)
Comment (ityuhui) [reply Changelog.md]: Updated. Thanks @leonardoalt.

Comment (chriseth) [reply Changelog.md]: Maybe we should be a bit more specific as to how it can be used? Something like `Allow `=` as separator between library name and address in --libraries commandline option`?

Comment (ityuhui) [reply Changelog.md]: Thanks @chriseth, updated.

File: solc/CommandLineInterface.cpp (hunk 6 above)
Comment (chriseth) [reply solc/CommandLineInterface.cpp]: This is the only place where an option `--libraries "x==0x1235"` is caught, right? In that case, it should not be a `solAssert` but instaed an error message using `serr` like above.

File: solc/CommandLineInterface.cpp [hunk 7]
```diff
@@ -664,34 +664,51 @@ bool CommandLineInterface::parseLibraryOption(string const& _input)
...
 
-			string libName(lib.begin(), lib.begin() + static_cast<ptrdiff_t>(colon));
+			string libName(lib.begin(), lib.begin() + static_cast<ptrdiff_t>(separator));
```
Comment (chriseth) [inline solc/CommandLineInterface.cpp:691]:
```suggestion
			string libName = lib.substr(0, separator);
```
?

Comment (chriseth) [reply solc/CommandLineInterface.cpp:691]: Ah, it was like that before.

File: docs/using-the-compiler.rst [hunk 8]
```diff
@@ -74,7 +74,7 @@ identify which libraries the placeholders represent. Note that the fully qualifi
 is the path of its source file and the library name separated by ``:``.
 You can use ``solc`` as a linker meaning that it will insert the library addresses for you at those points:
 
-Either add ``--libraries "file.sol:Math:0x1234567890123456789012345678901234567890 file.sol:Heap:0xabCD567890123456789012345678901234567890"`` to your command to provide an address for each library (use commas or spaces as separators) or store the string in a file (one library per line) and run ``solc`` using ``--libraries fileName``.
+Either add ``--libraries "file.sol:Math=0x1234567890123456789012345678901234567890 file.sol:Heap=0xabCD567890123456789012345678901234567890"`` to your command to provide an address for each library (use commas or spaces as separators) or store the string in a file (one library per line) and run ``solc`` using ``--libraries fileName``.
```
Comment (axic) [inline docs/using-the-compiler.rst:78]: I'd add
```
... note:
    Starting Solidity 0.8.1 accepts ``=`` as separator between library and address, and ``:`` as a separator is deprecated. It will be removed in the future. Currently `--libraries "file.sol:Math:0x1234567890123456789012345678901234567890 file.sol:Heap:0xabCD567890123456789012345678901234567890"`` will work too.
```

File: solc/CommandLineInterface.cpp [hunk 9]
```diff
@@ -664,34 +664,51 @@ bool CommandLineInterface::parseLibraryOption(string const& _input)
...
+
 			if (addrString.substr(0, 2) == "0x")
 				addrString = addrString.substr(2);
-			if (addrString.empty())
+			else
 			{
-				serr() << "Empty address provided for library \"" << libName << "\":" << endl;
-				serr() << "Note that there should not be any whitespace after the colon." << endl;
+				serr() << "The address " << addrString << " is not prefixed with \"0x\"." << endl;
+				serr() << "Note that the address should be prefixed with \"0x\"." << endl;
```
Comment (axic) [inline solc/CommandLineInterface.cpp]:
```suggestion
				serr() << "Note that the address must be prefixed with \"0x\"." << endl;
```

File: Changelog.md [hunk 10]
```diff
@@ -13,6 +13,7 @@ Compiler Features:
  * SMTChecker: Support try/catch statements.
  * SMTChecker: Output internal and trusted external function calls in a counterexample's transaction trace.
  * SMTChecker: Synthesize untrusted functions called externally.
+ * Command Line Interface: Allow "=" as separator between library name and address in ``--libraries`` commandline option.
```
Comment (axic) [inline Changelog.md:22]: Apparently we are also dropping support for non-0x prefix? If so, we should mention it here.

File: solc/CommandLineInterface.cpp (hunk 7 above)
Comment (ityuhui) [reply solc/CommandLineInterface.cpp:691]: I don't understand "like that before" because my change 
```c++
string libName(lib.begin(), lib.begin() + static_cast<ptrdiff_t>(separator));
```
is like before.  Just change "colon" to "separator"

File: docs/using-the-compiler.rst (hunk 8 above)
Comment (ityuhui) [reply docs/using-the-compiler.rst:78]: I don't know the format of ``note``, but I see other "note" or "warning" in this file (using-the-compiler.rst) is prefixed with two dots, and suffixed with two colons, e.g.

``.. warning::``

So could you please check whether my update is right ?

File: solc/CommandLineInterface.cpp (hunk 6 above)
Comment (ityuhui) [reply solc/CommandLineInterface.cpp]: OK. Done

File: solc/CommandLineInterface.cpp (hunk 9 above)
Comment (ityuhui) [reply solc/CommandLineInterface.cpp]: Done.

File: Changelog.md (hunk 10 above)
Comment (ityuhui) [reply Changelog.md:22]: OK. Added.

File: docs/using-the-compiler.rst [hunk 11]
```diff
@@ -74,7 +74,10 @@ identify which libraries the placeholders represent. Note that the fully qualifi
...
+
+... note:
```
Comment (leonardoalt) [inline docs/using-the-compiler.rst]:
```suggestion
.. note::

```
I think it should be like this

File: docs/using-the-compiler.rst (hunk 8 above)
Comment (axic) [reply docs/using-the-compiler.rst:78]: Yes, `.. note::` is the correct one.

File: docs/using-the-compiler.rst (hunk 11 above)
Comment (ityuhui) [reply docs/using-the-compiler.rst]: Thanks @leonardoalt

File: docs/using-the-compiler.rst (hunk 8 above)
Comment (ityuhui) [reply docs/using-the-compiler.rst:78]: Updated.

File: solc/CommandLineInterface.cpp (hunk 7 above)
Comment (ityuhui) [reply solc/CommandLineInterface.cpp:691]: Hi @chriseth 

Should I update this line to
```
string libName = lib.substr(0, separator);
```

**Review by chriseth — APPROVED** [review]

**Comment by chriseth** [comment]
Thanks a lot for your help!
